File: ui_test/prompts.py
"""Central prompt registry loader.

All LLM prompts live in prompts.yaml at the repo root. Code fetches them by
dotted key and substitutes {{name}} placeholders — never hardcode prompt text.
"""

from __future__ import annotations

import re
from functools import lru_cache
from pathlib import Path
from typing import Any

import yaml

PROMPTS_PATH = Path(__file__).resolve().parents[1] / "prompts.yaml"

_PLACEHOLDER = re.compile(r"\{\{(\w+)\}\}")


@lru_cache(maxsize=1)
def _registry() -> dict[str, Any]:
    data = yaml.safe_load(PROMPTS_PATH.read_text(encoding="utf-8"))
    if not isinstance(data, dict):
        raise ValueError(f"prompts.yaml did not parse to a mapping: {PROMPTS_PATH}")
    return data


def get_prompt(key: str) -> str:
    """Fetch a prompt by dotted key, e.g. 'exploration.decide'."""
    node: Any = _registry()
    for part in key.split("."):
        if not isinstance(node, dict) or part not in node:
            raise KeyError(f"Prompt '{key}' not found in {PROMPTS_PATH}")
        node = node[part]
    if not isinstance(node, str):
        raise KeyError(f"Prompt '{key}' is not a string in {PROMPTS_PATH}")
    return node.strip()


def render_prompt(key: str, **variables: str) -> str:
    """Fetch a prompt and substitute {{name}} placeholders."""
    template = get_prompt(key)

    def repl(match: re.Match[str]) -> str:
        name = match.group(1)
        if name not in variables:
            raise KeyError(f"Prompt '{key}' placeholder '{{{{{name}}}}}' has no value")
        return str(variables[name])

    return _PLACEHOLDER.sub(repl, template)

File: test-runner/server/prompts.ts
/**
 * Central prompt registry loader.
 *
 * All LLM prompts live in prompts.yaml at the repo root. Code fetches them by
 * dotted key and substitutes {{name}} placeholders — never hardcode prompt text.
 */

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { parse as parseYaml } from "yaml";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const PROMPTS_PATH = path.join(__dirname, "../../prompts.yaml");

let registry: Record<string, unknown> | null = null;

function loadRegistry(): Record<string, unknown> {
  if (registry) return registry;
  const raw = fs.readFileSync(PROMPTS_PATH, "utf8");
  const parsed = parseYaml(raw) as Record<string, unknown>;
  if (!parsed || typeof parsed !== "object") {
    throw new Error(`prompts.yaml did not parse to a mapping: ${PROMPTS_PATH}`);
  }
  registry = parsed;
  return registry;
}

/** Fetch a prompt by dotted key, e.g. "collaboration.helper_system". */
export function getPrompt(key: string): string {
  let node: unknown = loadRegistry();
  for (const part of key.split(".")) {
    if (!node || typeof node !== "object" || !(part in (node as Record<string, unknown>))) {
      throw new Error(`Prompt '${key}' not found in ${PROMPTS_PATH}`);
    }
    node = (node as Record<string, unknown>)[part];
  }
  if (typeof node !== "string") {
    throw new Error(`Prompt '${key}' is not a string in ${PROMPTS_PATH}`);
  }
  return node.trim();
}

/** Fetch a prompt and substitute {{name}} placeholders. */
export function renderPrompt(key: string, variables: Record<string, string>): string {
  return getPrompt(key).replace(/\{\{(\w+)\}\}/g, (_match, name: string) => {
    if (!(name in variables)) {
      throw new Error(`Prompt '${key}' placeholder '{{${name}}}' has no value`);
    }
    return variables[name];
  });
}
